homeassistant/components/media_player/firetv.py

```diff
@@ -195,60 +155,134 @@ def source_list(self):
+                        self._state = STATE_STANDBY
+
+                    # Check for a wake lock (device is playing).
+                    elif self._firetv._wake_lock:
+                        self._state = STATE_PLAYING
+
+                    # Otherwise, device is paused.
+                    else:
+                        self._state = STATE_PAUSED
+
+                # Don't get the current app.
+                elif self._firetv._wake_lock:
+                    # Check for a wake lock (device is playing).
+                    self._state = STATE_PLAYING
+                else:
+                    # Assume the devices is on standby.
+                    self._state = STATE_STANDBY
+
+        except:  # pylint: disable=bare-except
+            _LOGGER.error('%s', 'Update encountered an exception; will ' +
+                          'attempt to re-establish the ADB connection in ' +
+                          'the next update')
+            self._firetv._adb = None
+
+    @adb_wrapper
     def turn_on(self):
         """Turn on the device."""
-        self._firetv.action('turn_on')
+        self._firetv.turn_on()
 
+    @adb_wrapper
     def turn_off(self):
         """Turn off the device."""
-        self._firetv.action('turn_off')
+        self._firetv.turn_off()
 
+    @adb_wrapper
     def media_play(self):
         """Send play command."""
-        self._firetv.action('media_play')
+        self._firetv.media_play()
 
+    @adb_wrapper
     def media_pause(self):
         """Send pause command."""
-        self._firetv.action('media_pause')
+        self._firetv.media_pause()
 
+    @adb_wrapper
     def media_play_pause(self):
         """Send play/pause command."""
-        self._firetv.action('media_play_pause')
+        self._firetv.media_play_pause()
+
+    @adb_wrapper
+    def media_stop(self):
+        """Send stop (back) command."""
+        self._firetv.back()
 
+    @adb_wrapper
     def volume_up(self):
         """Send volume up command."""
-        self._firetv.action('volume_up')
+        self._firetv.volume_up()
 
+    @adb_wrapper
     def volume_down(self):
         """Send volume down command."""
-        self._firetv.action('volume_down')
+        self._firetv.volume_down()
 
+    @adb_wrapper
     def media_previous_track(self):
         """Send previous track command (results in rewind)."""
-        self._firetv.action('media_previous')
+        self._firetv.media_previous()
 
+    @adb_wrapper
     def media_next_track(self):
         """Send next track command (results in fast-forward)."""
-        self._firetv.action('media_next')
+        self._firetv.media_next()
 
+    @adb_wrapper
     def select_source(self, source):
         """Select input source."""
-        self._firetv.start_app(source)
+        if isinstance(source, str):
+            if not source.startswith('!'):
+                self._firetv.launch_app(source)
+            else:
+                self._firetv.stop_app(source[1:].lstrip())
```

Comment:
Mind documenting what is going on here?

Reply:
For example, selecting the source `com.netflix.ninja` will launch Netflix, while selecting the source`!com.netflix.ninja` will close Netflix.